### PR TITLE
Support Automated Highly Available Redis Deployments

### PR DESCRIPTION
**Pull request by kierenevans**
Automate a previously manual process by deploying the Highly Available Redis + Sentinels in stages.

### CODE REVIEW

File: redis/3-highly-available/README.md [hunk 1]
```diff
@@ -33,3 +33,158 @@ The following variables are supported
...
+```yaml
+variables:
+  REDIS_REPLICAS: 3
+  REDIS_SENTINEL_REPLICAS: 3
+
+tasks:
+  infrastructure:
+    deploy:
+      services:
+        mysql: ~
```
Comment (sroze) [inline redis/3-highly-available/README.md]: Any reason to include MySQL in this example? 🤔

File: redis/3-highly-available/README.md [hunk 2]
```diff
@@ -33,3 +33,158 @@ The following variables are supported
...
+    filter:
+      expression: 'tasks.infrastructure.services.mysql.created'
+    deploy:
+      services:
+        redis_initial_master:
+          specification:
+            source:
+              image: quay.io/continuouspipe/redis3-highly-available:stable
+            accessibility:
+              from_cluster: true
```
Comment (sroze) [inline redis/3-highly-available/README.md]: That's the default value so you can remove

File: redis/3-highly-available/README.md [hunk 3]
```diff
@@ -33,3 +33,158 @@ The following variables are supported
...
+            ports:
+              - 6379
+              - 26379
+            resources:
+              requests:
+                cpu: 50m
+                memory: 100Mi
```
Comment (sroze) [inline redis/3-highly-available/README.md]: For such memory-focused software, I'd put memory requests = memory limits

File: redis/3-highly-available/README.md [hunk 4]
```diff
@@ -33,3 +33,158 @@ The following variables are supported
...
+              port: 6379
+
+        redis_sentinel:
+          specification:
+            scalability:
+              number_of_replicas: ${REDIS_SENTINEL_REPLICAS}
+            source:
+              image: quay.io/continuouspipe/redis3-highly-available:stable
+            accessibility:
+              from_cluster: true
```
Comment (sroze) [inline redis/3-highly-available/README.md]: Same here re the default value & resources

File: redis/3-highly-available/README.md [hunk 5]
```diff
@@ -33,3 +33,158 @@ The following variables are supported
...
+              requests:
+                cpu: 50m
+                memory: 100Mi
+              limits:
+                cpu: 250m
+                memory: 500Mi
+            command:
+              - /bin/bash
+              - -c
+              - '/usr/local/bin/redis_command run_master_and_sentinel'
```
Comment (sroze) [inline redis/3-highly-available/README.md]: Why don't we call it `container` as the rest?

File: redis/3-highly-available/usr/local/share/redis_functions.sh [hunk 6]
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
+
+function get_existing_master()
+{
+  redis-cli -h redis-sentinel -p 26379 --csv SENTINEL get-master-addr-by-name mymaster | tr ',' ' ' | cut -d' ' -f1 | cut -d'"' -f2
```
Comment (sroze) [inline redis/3-highly-available/usr/local/share/redis_functions.sh]: maybe you can abstract `redis-cli -h redis-sentinel -p 26379` in a function?

File: redis/3-highly-available/usr/local/share/redis_functions.sh [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+function sentinel_cleanup()
+{
+  local EXISTING_MASTER="$1"
+  # Run reset on each sentinel.
+  SENTINELS="$(echo -e "SENTINEL sentinels mymaster\nSENTINEL RESET mymaster\n" | redis-cli -h redis-sentinel -p 26379 --csv | sed 's/"name"/\n"name"/g' | grep "name" | grep -v 's-down-time' | sed -E 's/.*"ip","([^"]+)".*/\1/')"
+  if [ -n "$EXISTING_MASTER" ]; then
+    SENTINELS="$(echo "$SENTINELS" | grep -v "$EXISTING_MASTER")"
+  fi
+  for sentinel in $SENTINELS; do
+    sleep 11s
```
Comment (sroze) [inline redis/3-highly-available/usr/local/share/redis_functions.sh]: where does that come from? :-)

File: redis/3-highly-available/README.md (hunk 1 above)
Comment (kierenevans) [reply redis/3-highly-available/README.md]: Good point, it's to ensure it only happens the first time an environment is created.

Comment (kierenevans) [reply redis/3-highly-available/README.md]: Is there a better way to do that?

File: redis/3-highly-available/README.md (hunk 2 above)
Comment (kierenevans) [reply redis/3-highly-available/README.md]: Done!

File: redis/3-highly-available/README.md (hunk 3 above)
Comment (kierenevans) [reply redis/3-highly-available/README.md]: Done!

File: redis/3-highly-available/README.md (hunk 5 above)
Comment (kierenevans) [reply redis/3-highly-available/README.md]: Done!

File: redis/3-highly-available/README.md (hunk 4 above)
Comment (kierenevans) [reply redis/3-highly-available/README.md]: Done!

File: redis/3-highly-available/usr/local/share/redis_functions.sh (hunk 7 above)
Comment (kierenevans) [reply redis/3-highly-available/usr/local/share/redis_functions.sh]: Added some explanation - though perhaps we can make it more objective than time-based, e.g. query if the previously reset sentinel has some other sentinel knowledge now. For another day perhaps.

File: redis/3-highly-available/usr/local/share/redis_functions.sh (hunk 6 above)
Comment (kierenevans) [reply redis/3-highly-available/usr/local/share/redis_functions.sh]: Done!

File: redis/3-highly-available/README.md (hunk 1 above)
Comment (sroze) [reply redis/3-highly-available/README.md]: I see. Maybe you can use a simple `nginx` container and add a comment to explain why? I read it as "Why the fuck Redis HA needs MySQL?" 😄

File: redis/3-highly-available/README.md [hunk 8]
```diff
@@ -33,3 +33,148 @@ The following variables are supported
...
+                memory: 500Mi
+
+  redis_ha_master_failover:
+    filter:
+      expression: 'tasks.infrastructure.services.mysql.created'
+    run:
+      image:
+        image: quay.io/continuouspipe/redis3-highly-available:stable
+      commands:
+        - /bin/bash -c '/usr/local/bin/container master_failover_and_sentinel_cleanup'
```
Comment (sroze) [inline redis/3-highly-available/README.md]: `container master_failover_and_sentinel_cleanup` doesn't work? 🤔

**Pull request merged**